### DISCUSSION TITLE
How do you eliminate echo and audio feedback?

### DISCUSSION THREAD

**Discussion by ygreq** [CESNET/UltraGrid#200] [Q&A]
Hi there! 

I would like to ask a few questions in regard to echo and audio feedback and squealing. I don't have much experience with sound as I have a mostly video and installation background), but I did search a bit about audio feedback and what is causing it. 

**The problem**
If I create an audio connection between 2 PCs via UltraGrid, I get audio feedback / squealing. And if I plug in a pair of headphones onto one of the PCs, I realize that I get echo of my voice from the other room (which does not use headphones) which is annoying for the speaker to hear her/his own voice. This would probably be the reason for audio feedback in the end.   

For now I am trying to replicate a commercial audio chat application like FB Messenger or Skype because I would like to compare UG to them in relation to how they deal with echo and audio feedback. What I mean is **why, in a similar situation** (same PCs, same capture and playback devices, same spaces) **, FB Messenger and Skype would not produce echo and audio feedback, but Ultragrid does?** I know there are some circumstances where one can get echo or / and audio feedback while having a normal phone conversation and sometimes while using commercial chat apps. But still not often and not always like I do now with UG. :)

I am asking because maybe I did not choose the proper capture or playback devices or maybe I can add some extra code to it. 

**The setup**
I have 2 PCs in 2 different spaces running UltraGrid. The code is simple
PC1 `uv  -s portaudio:1 -r portaudio:4 [IP of PC2]`
PC2 `uv  -s portaudio:1 -r portaudio:4 [IP of PC1]`

The capture device for both PCs is portaudio:1 - Microphone (Realtek(R) Audio) (output channels: 0; input channels: 2; MME) while the playback device is portaudio:4 - Speakers (Realtek(R) Audio) (output channels: 2; input channels: 0; MME)

**Observations**
Most of tips were related to stage setup and similar where you can manage the problem by positioning the microphones and the speakers properly. But for digital communication via chat apps (with internal devices) I couldn't find many solutions. My guess is that these problems which seem to be normal are solved in the software itself. Maybe I am wrong. 

Please let me know what you think and how to fix this. 
Thank you!

**Comment by MartinPulec** [comment]
This is quite simple - UltraGrid doesn't currently incorporate an echo-canceller. There is actually one in UG, but it is currently disabled, we may fix it in future. The difference is that commercial chat application solve easier problem - it usually uses single sound device with the same input/output clock. This is not necessarily true for UltraGrid where you may have completely independent device, like DeckLink on input and ALSA on output.

If using Linux with PulseAudio, you may also try its [echo canceller](https://github.com/baresip/baresip/wiki/Using-PulseAudio:-module-echo-cancel), which would work below UltraGrid.

**Reply by ygreq** [reply]
Hi, Martin! Thank you so much for your advice in this matter. I don't believe I get all you are saying regarding what UG can do. 

I use the apps on Windows so, unfortunately, I can't use the app in the link. So **I am using UG in Windows and using portaudio for both capture and playback devices** with actual devices being internal microphones of laptops (or external microphones) and internal speakers of laptops (or external speakers). The tests I am doing now use internal physical devices. 

In my situation, **you mean to say I am stuck with echo and audio feedback?** :/

**Comment by MartinPulec** [comment]
I don't have own experience but you can check [this tools](https://krisp.ai/blog/noise-cancelling-apps-windows/), I believe that it may do the same job as Pulseaudio. Although the headline of the list says noise cancelling, right the first – Krisp – states that it does echo cancellation as well.

**Reply by ygreq** [reply]
Thank you, Martin. I will try them. 

I am curious. How would you recommend UltraGrid be used in order not to have echo? I mean what type of additional software and what hardware needs to be used in order to eliminate echo? You mentioned this: <<This is not necessarily true for UltraGrid where you may have completely independent device, like DeckLink on input and ALSA on output.>> May I ask that you tell me more? For example if I use DeckLink on input and ALSA on output on both ends, I will eliminate echo? What other combinations are there for Windows machines?

Thank you so much!

**Comment by mpiatka** [comment]
Hi! I've done some work on the echo canceller that is included in UltraGrid. To try it you can download a [build from here.](https://github.com/mpiatka/UltraGrid/suites/4173980312/artifacts/107396891) To enable it, pass the argument `--echo-cancellation`

Note, that it is currently pretty experimental and there are tons of limitations. Firstly, the sampling clock of the recording and playback devices must be the same (using the internal mic and speakers should be fine). Second, the sound system should not have any nonlinearities or distortions (no clipping must be present, also various "sound enhancements" or equalizers could impair echo cancellation). Finally, the echo canceller is very sensitive to audio drop outs, so there should be no packet drops on the network.

An external echo canceller (like Martin mentioned earlier) could provide a better user experience if you can get it working.

**Reply by ygreq** [reply]
Thank you, mpiatka. I will test your solution asap. 

Related to this << the echo canceller is very sensitive to audio drop outs, so there should be no packet drops on the network.>> is that even possible? No packet drops, I mean.

**Reply by mpiatka** [reply]
The echo canceller needs the recording and playback streams to be synchronized as close as possible and it also needs a little time to "calibrate" itself before it starts working. The problem is, that in UltraGrid it is currently quite hard to properly synchronize the streams for the echo canceller, which means that every time there is an audio underflow the echo canceller needs to calibrate again.

A few very rare packet drops will perhaps not be too much of a problem, but every time the audio drops out, the canceller will stop working for a few seconds.

**Reply by ygreq** [reply]
I understand. Thank you so much, mpiatka, for these thoughtful explanations.
